src/timed-stream.ts

```diff
@@ -0,0 +1,113 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {PassThrough, TransformCallback, TransformOptions} from 'stream';
+
+/**
+ * This interface is the usual options that can be passed into a Transform plus
+ * a hook for injecting code into the stream's transform function. This hook is
+ * useful for code running code that normally runs in the transform method if that
+ * code is different for each method that makes use of a stream.
+ */
+type TimedStreamOptions = TransformOptions & {
+  transformHook?: (
+    event: any,
+    _encoding: BufferEncoding,
+    callback: TransformCallback,
+  ) => void;
+};
+
+class StreamTimer {
+  private startTime;
+  private totalDuration;
+
+  constructor() {
+    this.startTime = 0n;
+    this.totalDuration = 0n;
+  }
+
+  getTotalDurationMs() {
+    return Number(this.totalDuration / 1_000_000n);
+  }
+
+  start() {
+    this.startTime = process.hrtime.bigint();
+  }
+
+  stop() {
+    const endTime = process.hrtime.bigint();
+    const duration = endTime - this.startTime;
+    this.totalDuration += duration;
+  }
+}
+
+/**
+ * The TimedStream class is used for measuring the time the user spends
+ * processing data from the stream. We need to measure this time for use cases
+ * like measuring the application latencies for client side metrics.
+ */
+export class TimedStream extends PassThrough {
+  private readTimer = new StreamTimer();
+  private transformTimer = new StreamTimer();
+  constructor(options?: TimedStreamOptions) {
+    // highWaterMark of 1 is needed to respond to each row
```

Comment:
is this still relevant? The code itself is using highWaterMark of 0

Reply:
I'll address this in a [backlog bug](https://b.corp.google.com/issues/433566793) because this is merge ready.